docs: deploy Eliza to TEE with Oasis ROFL app (#4334)
elizaOS/eliza: docs/docs/advanced/deploy-eliza-to-tee-with-oasis-rofl.md

        build:
          context: .
          dockerfile: Dockerfile
        platform: linux/amd64
        environment:
          - CHARACTERS=${CHARACTERS}
          - SERVER_PORT=${SERVER_PORT}
          # Add other required environment variables
    ```

2. **Encrypt Secrets**

    Encrypt all required environment variables. You can use any character file of your choice - here we're using the staff engineer character as an example. Refer to the Secret Management section above for a comprehensive list of secrets you might want to set:

    ```bash
    # Set the character file path (you can use any character file of your choice)
    echo -n "characters/staff-engineer.character.json" | oasis rofl secret set CHARACTERS -
    echo -n "3000" | oasis rofl secret set SERVER_PORT -
    # Add other required secrets
    ```

3. **Build the ROFL Application**

    For MacOS users, use the following Docker image:

    ```bash
    docker run --platform linux/amd64 --volume .:/src --rm -it ghcr.io/oasisprotocol/rofl-dev:main
    ```

    Inside the container, build the ROFL application:

    ```bash
    oasis rofl build --update-manifest
    ```

4. **Deploy the ROFL Application**

>     You have two options for deployment:

Reviewer: This step has changed slightly from the last month. The Oasis CLI deployment is implemented in https://github.com/oasisprotocol/cli/pull/422 and it should be the preferred way for people to deploy .orc files—either to their own nodes (you are the provider), Oasis-sponsored nodes or 3rd party nodes. The PR should be merged in the upcoming days.

The basic command to deploy your orc is `oasis rofl deploy`. If you don't have an existing machine available, you are asked to pick a provider+the desired plan to rent a TDX-capable machine (and pay for it).

    A. **Run Your Own Oasis Node**

    1. Follow the [Oasis Node Setup Guide](https://docs.oasis.io/node/run-your-node/paratime-client-node)
    2. Copy the ROFL bundle to your node
    3. Update your node configuration
    4. Restart your node

    B. **Deploy to Oasis Provider**

    1. Upload the ROFL bundle to a public file server
    2. Contact the Oasis team on [Discord](https://oasis.io/discord) #dev-central channel

### Verify TEE Deployment

> Once deployed, you can verify the TEE deployment by:
> 
> 1. Checking the ROFL application status:
> 
>     ```bash
>     oasis rofl show
>     ```
> 
> 2. Monitoring the application logs
> 3. Verifying remote attestation

Reviewer: Part of the Oasis CLI PR above is also this.

````suggestion
Once deployed, you can verify whether your ROFL instance is running, the latest logs and the runtime-attestation key:

    ```bash
    oasis rofl machine show
    ```
````

![](/img/eliza_in_tee.jpg)

## Overview

Eliza can be deployed to a Trusted Execution Environment (TEE) using the [Oasis SDK](https://github.com/oasisprotocol/oasis-sdk). This guide will walk you through the process of setting up and running Eliza in a TEE environment using the Oasis ROFL tool.

### Background

The Oasis ROFL tool is built on top of the [Oasis SDK](https://github.com/oasisprotocol/oasis-sdk), which is designed to simplify the deployment of applications to TEE environments. The main features include:

- Secure deployment of applications to TEE environments
- Remote attestation and verification
- Support for both development and production environments
- Integration with the Oasis network

---

## Core Components

Eliza's TEE implementation consists of several key components that work together to provide a secure execution environment:

1. ROFL Application Configuration
2. Secret Management
3. Build and Deployment System

These components ensure:
- Secure execution of the Eliza agent
- Protected handling of sensitive data
- Verifiable proof of TEE execution
- Support for both development and production environments

---

### ROFL Application Configuration

> The ROFL application configuration handles the setup and management of the TEE environment. It supports:
> 
> - Multiple deployment modes:
>     - `testnet`: For development and testing
>     - `mainnet`: For production deployment
> 
> Key features:

Reviewer: Leftover?

```suggestion
The ROFL application configuration handles the setup and management of the TEE environment. It supports:

- Multiple deployment modes:
    - `testnet`: For development and testing
    - `mainnet`: For production deployment
```

> - Secure application deployment
> - Configuration management

Reviewer: Configuration management is probably deployment mode above?
Also I suggest you mention Docker compose below to ring a bell for those who know it.

```suggestion
- Secure application deployment
- Raw or Docker compose-based container management
```

- Policy enforcement
- Resource allocation

Example configuration:

```yaml
services:
  eliza:
    restart: always
    image: ghcr.io/elizaos/eliza:latest
    build:
      context: .
      dockerfile: Dockerfile
    platform: linux/amd64
    environment:
      - CHARACTERS=${CHARACTERS}
      - SERVER_PORT=${SERVER_PORT}
```

---

### Secret Management

> The secret management system handles secure storage and access to sensitive data. It:

Reviewer: ```suggestion
The ROFL secret management system handles secure storage and access to sensitive data. It:
```

> - Encrypts environment variables and secrets

Reviewer: ```suggestion
- Encrypts secrets and stores them on-chain
```

- Provides secure access to sensitive data
- Supports both development and production environments

> Key features:
> - Secure secret storage
> - Environment variable encryption
> - Access control
> - Audit logging
> 

Reviewer: This is redundant IMO. Also, access control is not supported. Audit logging is only implicit—when ROFL is spun up.

```suggestion
```

echo -n "" | oasis rofl secret set MEDIUM_OPENAI_MODEL -
echo -n "" | oasis rofl secret set LARGE_OPENAI_MODEL -
echo -n "" | oasis rofl secret set EMBEDDING_OPENAI_MODEL -
echo -n "" | oasis rofl secret set IMAGE_OPENAI_MODEL -
echo -n "false" | oasis rofl secret set USE_OPENAI_EMBEDDING -

# GitHub Integration
echo -n "true" | oasis rofl secret set GITHUB_CLIENT_DISABLED -
echo -n "true" | oasis rofl secret set GITHUB_PLUGIN_ENABLED -
echo -n "ghp_XXXX" | oasis rofl secret set GITHUB_API_TOKEN -
echo -n "5000" | oasis rofl secret set GITHUB_USER_CHECK_INTERVAL_MS -
echo -n "5000" | oasis rofl secret set GITHUB_INFO_DISCOVERY_INTERVAL_MS -
echo -n "20000" | oasis rofl secret set GITHUB_OODA_INTERVAL_MS -
echo -n "10" | oasis rofl secret set GITHUB_ISSUES_LIMIT -
echo -n "10" | oasis rofl secret set GITHUB_PULL_REQUESTS_LIMIT -

# Database Configuration
echo -n "postgresql://user:password@localhost:5432/eliza" | oasis rofl secret set POSTGRES_URL -
```

## Tutorial

---

### Prerequisites

Before getting started with Eliza in TEE, ensure you have:

- [Oasis CLI](https://github.com/oasisprotocol/cli/releases) installed
- Docker Desktop or Orbstack (Orbstack is recommended)

> - Testnet tokens from the [Oasis Faucet](https://faucet.testnet.oasis.dev/)

Reviewer: This is redundant here IMO, because you probably don't have a keypair yet which is described below.

```suggestion
```

- A character file for your Eliza agent (you can use the example character file or create your own)

---

### Environment Setup

To set up your environment for TEE development:

1. **Create a keypair**

    ```bash
    oasis wallet create <account-name>
    ```

> 2. **Request Testnet Tokens**

Reviewer: ```suggestion
2. **Request TEST Tokens**
```

>     Visit the [Oasis Faucet](https://faucet.testnet.oasis.dev/) to request TEST tokens.

Reviewer: ```suggestion
    Visit the [Oasis Testnet Faucet](https://faucet.testnet.oasis.dev/) to request TEST tokens to the address corresponding to your keypair.
```

3. **Create a ROFL Application**

>     Create the ROFL application:

Reviewer: ```suggestion
    Create the ROFL application on Testnet:
```

    ```bash

>     oasis rofl create --network testnet --account <account-name>

Reviewer: This will require an existing `rofl.yaml` in the project. Is it included? Otherwise one should run `oasis rofl init` in their project directory first.

    ```

### Build and Deploy the ROFL app

1. **Configure the ROFL app**

>     Update the `rofl-compose.yaml` file with your configuration:

Reviewer: By default `oasis rofl init` will create `compose.yaml` (and reference it in `rofl.yaml`). `rofl-compose.yaml` is preferred naming, when the existing project already uses Docker containers and the container configuration conflicts with ROFL.

    echo -n "characters/staff-engineer.character.json" | oasis rofl secret set CHARACTERS -
    echo -n "3000" | oasis rofl secret set SERVER_PORT -
    # Add other required secrets
    ```

    **WARNING:** Secrets are end-to-end encrypted with your account key and an ephemeral key of the chain where it will be deployed to (Testnet or Mainnet). Migrating encrypted secrets from one deployment to another is not possible.

3. **Configure ROFL Resources**

    Adjust the resource allocation in your `rofl.yaml` file to match your application's requirements. This includes memory, CPU, and storage settings:

    ```yaml
    resources:
      memory: 16384  # Memory in MB (16GB in this example)
      cpus: 1        # Number of CPU cores
      storage:
        kind: disk-persistent
        size: 10240  # Storage size in MB (10GB in this example)
    ```

    Adjust these values based on your specific needs:
    - `memory`: Set the amount of RAM your application needs (in MB)
    - `cpus`: Specify the number of CPU cores to allocate
    - `storage`: Configure persistent storage with the required size

4. **Build the ROFL Application**

>     For MacOS users, use the following Docker image:

Reviewer: Anyone can use Docker to build ROFL. In fact, it's the preferred way to do it if you need reproducible builds.

```suggestion
    Use the following Docker image for reproducible ROFL builds:
```

    ```bash
    docker run --platform linux/amd64 --volume .:/src --rm -it ghcr.io/oasisprotocol/rofl-dev:main oasis rofl build
    ```

5. **Deploy the ROFL Application**

    The recommended way to deploy your ROFL application is using the Oasis CLI:

    ```bash
    oasis rofl deploy
    ```

    This command will:

>     1. Check if you have an existing machine available

Reviewer: ```suggestion
    1. Check if you have an existing ROFL provider selected and machine available
```

    2. If no machine exists, it will:
       - Present a list of available providers
       - Allow you to select a provider and plan for a TDX-capable machine

>        - Handle the payment process for renting the machine

Reviewer: ```suggestion
       - Handle the payment process for renting the machine (Testnet machines can be paid with TEST tokens)
```

    3. Deploy your ROFL application to the selected machine

    You can deploy to:

>     - Your own nodes (if you're the provider)

Reviewer: I would remove mentioning being provider, because anyone can be a provider. There is no whitelisting or registration at Oasis foundation needed.

```suggestion
    - Your own nodes
```

>     - Oasis-sponsored nodes

Reviewer: Sponsored is maybe too generous. For Mainnet, people will still need to pay ROSE (and/or other currencies) for renting the machine.

```suggestion
    - Nodes managed by the Oasis Protocol Foundation
```

    - Third-party nodes

>     For more advanced deployment options and configuration, refer to the [Oasis CLI documentation](https://docs.oasis.io/cli/).

Reviewer: ```suggestion
    For more advanced deployment options and configuration, refer to the [official ROFL deployment documentation](https://docs.oasis.io/build/rofl/deployment).
```

### Background

The Oasis ROFL tool is built on top of the [Oasis SDK](https://github.com/oasisprotocol/oasis-sdk), which is designed to simplify the deployment of applications to TEE environments. The main features include:

- Secure deployment of applications to TEE environments
- Remote attestation and verification
- Support for both development and production environments
- Integration with the Oasis network

---

## Core Components

Eliza's TEE implementation consists of several key components that work together to provide a secure execution environment:

1. ROFL Application Configuration
2. Secret Management
3. Build and Deployment System

These components ensure:
- Secure execution of the Eliza agent
- Protected handling of sensitive data
- Verifiable proof of TEE execution
- Support for both development and production environments

---

### ROFL Application Configuration

The ROFL application configuration (`rofl.yaml`) defines how your application will be deployed and executed in the TEE environment. The configuration supports different deployment modes:

1. **Self-hosted Deployment** (when you run your own node)

> 2. **Oasis-sponsored Deployment** (using Oasis-provided infrastructure)

Reviewer: ```suggestion
2. **Oasis-provided Deployment** (using nodes maintained by the Oasis Protocol Foundation)
```

3. **Third-party Deployment** (using other provider's infrastructure)

> Here's a comprehensive example configuration:

Reviewer: ```suggestion
For illustration, here's a comprehensive example configuration:
```

> ```yaml

Reviewer: ```suggestion
```yaml title="rofl.yaml"
```

name: eliza
version: 0.11.1
tee: tdx
kind: container
resources:
  memory: 16384  # Memory in MB (16GB)
  cpus: 1        # Number of CPU cores
  storage:
    kind: disk-persistent

>     size: 10240  # Storage size in MB (10GB)

Reviewer: No need for power of two.

```suggestion
    size: 10000  # Storage size in MB (10GB)
```

  kernel: https://github.com/oasisprotocol/oasis-boot/releases/download/v0.4.1/stage1.bin#[hash]
  stage2: https://github.com/oasisprotocol/oasis-boot/releases/download/v0.4.1/stage2-podman.tar.bz2#[hash]
  container:
    runtime: https://github.com/oasisprotocol/oasis-sdk/releases/download/rofl-containers%2Fv0.4.2/rofl-containers#[hash]
    compose: rofl-compose.yaml

# Deployment configurations
deployments:
  # Default deployment configuration
  default:
    app_id: [app_address]    # Your application's address
    network: [network]       # Network to deploy to (testnet/mainnet)
    paratime: [paratime]     # Paratime to use
    admin: [admin_address]   # Admin address for management

    # Trust root configuration
    trust_root:
      height: [height]       # Block height
      hash: [hash]          # Block hash

    # Security policy configuration
    policy:
      quotes:
        pcs:
          tcb_validity_period: 30
          min_tcb_evaluation_data_number: 18
          tdx: {}
      enclaves:
        - [enclave_1]        # Required enclave IDs
        - [enclave_2]
      endorsements:

>         - any: {}           # Endorsement policy

Reviewer: ```suggestion
        - any: {}           # Which nodes can run your ROFL
```

      fees: endorsing_node  # Fee payment method
      max_expiration: 3     # Maximum expiration time in days

    # Application secrets
    secrets:
      - name: [secret_name]
        value: [secret_value]
      # Add more secrets as needed

> ```

Reviewer: With Oasis CLI 0.13.0, the deployment also contains rented machines info.

````suggestion
    machines:
      default:
        provider: [provider_address] # Address of the provider hosting your ROFL
        offer: [offer_name] # Name of the rented plan
        id: [machine_id] # ID of the rented machine running your ROFL
```
````

> Key configuration points:

Reviewer: ```suggestion
The configuration file above is generated and managed by the [Oasis CLI]. To better understand the bells and whistles behind ROFL however, let's look at the key configuration fields:
```

- `resources`: Adjust based on your application's needs
- `artifacts`: Required for TDX execution - do not modify unless you know what you're doing
- `deployments`: Define different deployment configurations for various environments (Raw or Docker compose-based container)
- `policy`: Configure security requirements and attestation policies
- `secrets`: Manage your application's sensitive data

> 

Reviewer: ```suggestion
- `machines`: Configuration of the selected provider and rented machines

```

Reviewer 2: @snobbee please address this going to resolve for now so we can get this merged in

echo -n "" | oasis rofl secret set OPENAI_API_URL -
echo -n "" | oasis rofl secret set SMALL_OPENAI_MODEL -
echo -n "" | oasis rofl secret set MEDIUM_OPENAI_MODEL -
echo -n "" | oasis rofl secret set LARGE_OPENAI_MODEL -
echo -n "" | oasis rofl secret set EMBEDDING_OPENAI_MODEL -
echo -n "" | oasis rofl secret set IMAGE_OPENAI_MODEL -
echo -n "false" | oasis rofl secret set USE_OPENAI_EMBEDDING -

# GitHub Integration
echo -n "true" | oasis rofl secret set GITHUB_CLIENT_DISABLED -
echo -n "true" | oasis rofl secret set GITHUB_PLUGIN_ENABLED -
echo -n "ghp_XXXX" | oasis rofl secret set GITHUB_API_TOKEN -
echo -n "5000" | oasis rofl secret set GITHUB_USER_CHECK_INTERVAL_MS -
echo -n "5000" | oasis rofl secret set GITHUB_INFO_DISCOVERY_INTERVAL_MS -
echo -n "20000" | oasis rofl secret set GITHUB_OODA_INTERVAL_MS -
echo -n "10" | oasis rofl secret set GITHUB_ISSUES_LIMIT -
echo -n "10" | oasis rofl secret set GITHUB_PULL_REQUESTS_LIMIT -

# Database Configuration
echo -n "postgresql://user:password@localhost:5432/eliza" | oasis rofl secret set POSTGRES_URL -
```

## Tutorial

---

### Prerequisites

Before getting started with Eliza in TEE, ensure you have:

> - [Oasis CLI](https://github.com/oasisprotocol/cli/releases) installed

Reviewer: ```suggestion
- [Oasis CLI] installed
```

- Docker Desktop or Orbstack (Orbstack is recommended)
- Testnet tokens from the [Oasis Faucet](https://faucet.testnet.oasis.dev/)
- A character file for your Eliza agent (you can use the example character file or create your own)

> 

Reviewer: ```suggestion

[Oasis CLI]: https://github.com/oasisprotocol/cli/releases

```
